src/core/env.ml

```diff
@@ -1,79 +1,98 @@
-(** Scoping environment for variables. *)
+(** Scoping environment for variables.
+    Note that since meta-variable types are closed, the contexts passed to
+    {!val:Eval.whnf} and alike are empty. *)
 
+open Extra
 open Terms
 
 (** Type of an environment, used in scoping to associate names to
-   corresponding Bindlib variables and types. Note that it cannot be
-   implemented by a map as the order is important. *)
-type env = (string * (tvar * tbox)) list
+    corresponding Bindlib variables and types. Note that it cannot be
+    implemented by a map as the order is important. The boolean is used to
+    indicate whether the variable is defined in some context. *)
+type env = (string * (tvar * bool * tbox)) list
```

Comment:
Following my previous remark, we in fact need to take the following definition:
```
type env = (string * (tvar * tbox * tbox option)) list
```
where the first tbox is the type and the second one is the definition (I permuted the arguments to follow the style in other definitions).